components/clp-package-utils/clp_package_utils/general.py

```diff
@@ -581,21 +585,31 @@ def validate_dataset_exists(db_config: Database, dataset: str) -> None:
             raise ValueError(f"Dataset `{dataset}` doesn't exist.")
 
 
-def validate_dataset_name(dataset: str) -> None:
+def validate_dataset_name(clp_table_prefix: str, dataset_name: str) -> None:
     """
     Validates that the given dataset name abides by the following rules:
-    - Its length is between 1 and 44 characters.
+    - Its length won't cause any metadata table names to exceed MySQL's max table name length.
     - It only contains alphanumeric characters and underscores.
 
-    :param dataset:
+    :param clp_table_prefix:
+    :param dataset_name:
     :raise: ValueError if the dataset name is invalid.
     """
-    if re.fullmatch(r"\w+", dataset) is None:
+    if re.fullmatch(r"\w+", dataset_name) is None:
         raise ValueError(
-            f"Invalid dataset name: `{dataset}`. It must only contain alphanumeric characters and"
-            f" underscores."
+            f"Invalid dataset name: `{dataset_name}`. Names can only contain alphanumeric"
+            f" characters and underscores."
         )
-    if len(dataset) > 44:
+
+    num_underscores_in_table_name = 2
+    dataset_max_len = (
+        MYSQL_TABLE_NAME_MAX_LEN
+        - len(clp_table_prefix)
+        - TABLE_SUFFIX_MAX_LEN
+        - num_underscores_in_table_name
+    )
+    if len(dataset_name) > dataset_max_len:
         raise ValueError(
-            f"Invalid dataset name: `{dataset}`. It mustn't be longer than 44 characters."
+            f"Invalid dataset name: `{dataset_name}`. Names can only be a maximum of"
+            f" {dataset_max_len} characters long."
         )
```

Comment:
In general the code makes sense to me. just a nit.
1. iirc clp_table_prefix is defined to be `clp_` which already includes the under score. so the `num_underscores_in_table_name` should be 1 instead, and we probably need to rename it.


Unrelated to this PR, I also noticed that CLP_METADATA_TABLE_PREFIX seems to be not customizable by users.. wonder:
1. If we don't plan to let user changing it, can we always refer to the constant directly in our codebase?
2. Since we have all tables under clp_db, we should have full control of the table names, can we remove the prefix or will it cause any name conflict with mysql's default table?

Reply:
True, my mistake.

1. The long term goal was to let users change it so that if they really wanted, they could store clp's metadata in an existing database they had. However, how our metadata is evolving and things are becoming more performance critical, perhaps it is a good idea to get rid of it in another PR. Shall we ask the rabbit to open an issue?
2. Right, I guess technically we could get rid of the prefix from each table, but it might be better to leave it so the CLP tables are easily distinguishable from MySQL's system tables.